sql/log.cc

```diff
@@ -163,6 +163,111 @@ static SHOW_VAR binlog_status_vars_detail[]=
+
+    @retval true    it should do free flush
+    @retval false   it should do normal commit.
+  */
+  bool should_free_flush(const MYSQL_BIN_LOG::group_commit_entry *entry) const;
+
+  /**
+    This function is the entry function to do free flush. It first,
+    rotate the binlog, then rename the temporary file of the
+    binlog cache to new binlog file, after that it commits the transaction.
+
+    @param entry, group_commit_entry object of current transaction.
+
+    @retval true    free flush succeeds.
+    @retval false   free flush fails, it should go to normal commit process.
+  */
+  bool commit(MYSQL_BIN_LOG::group_commit_entry *entry);
+
+  /**
+    After rotate has created the new binlog file, it copies the content
+    of the new binlog file to the binlog cache, delete the new binlog file
+    and then rename the binlog cache to the new binlog file.
+
+    @retval true   Succeeds to replace the binlog file.
+    @retval false    Failed to replace the binlog file. It only return
+                    true if some error happened after the new binlog file
+                    is deleted. In this situation rotate process will fail.
+  */
+  bool replace_binlog_file();
+  /**
+    The space left is more than a gtid event required, thus the extra
+    space is padded into the gtid event as 0. This function is used
+    to calculate the real gtid size with pad.
+  */
+  size_t get_gtid_event_pad_size();
+
+  /**
+    The space required for session binlog caches to reserve. It is calculated
+    from the length of current binlog file when it is generated and aligned
+    to IO_SIZE;
+
+    @param header_len  header length of current binlog file.
+  */
+  void set_reserved_bytes(uint32 header_len)
+  {
+    // Gtid event length
+    header_len+= LOG_EVENT_HEADER_LEN + Gtid_log_event::max_data_length +
+                 BINLOG_CHECKSUM_LEN;
+    header_len= header_len - (header_len % IO_SIZE) + IO_SIZE;
+    if (header_len != m_reserved_bytes)
+      m_reserved_bytes= header_len;
+  }
+
+  /**
+    Return reserved space required for binlog cache. It is NOT defined as
+    an atomic variable, while it is get and set in parallel. Synchronizing
+    between set and get is not really necessary.
+  */
+  uint32 get_reserved_size()
+  {
+    return m_reserved_bytes;
+  }
+private:
+  Binlog_free_flush &operator=(const Binlog_free_flush &);
+  Binlog_free_flush(const Binlog_free_flush &);
+
+  char m_cache_dir[FN_REFLEN];
+
+  /** The commit entry of current transaction which is doing free flush. */
+  MYSQL_BIN_LOG::group_commit_entry *m_entry{nullptr};
+
+  /** The cache_data which will be renamed to binlog. */
+  binlog_cache_data *m_cache_data{nullptr};
+
+  /** It will be set to true if rename operation succeeds */
+  bool m_replaced{false};
+
+  uint32 m_reserved_bytes {IO_SIZE};
+};
+static Binlog_free_flush binlog_free_flush;
+ulonglong opt_binlog_free_flush_threshold= 10 * 1024 * 1024;
```

Comment:
I think this mismatches with the default `128 * 1024 * 1024` mentioned in `sys_vars.cc`